### ISSUE TITLE
Properties of \ArrayAccess objects may become inaccessible.

### ISSUE THREAD

**Issue by kitsunet** [TYPO3Fluid/Fluid#184]
The following scenario is currently not solvable with the Accessor handling as is, but was before standalone:

Given a class roughly like this (skipped some implementation details):
```
class Foo implements \ArrayAccess
{
  // ArrayAccess implementation

  getSomePropertyNotAccessibleByArrayAccess()
  {
    return $this->somePropertyNotAccessibleByArrayAccess;
  }
}
```

Instance available in the template as `foo`:

```
{foo.somePropertyNotAccessibleByArrayAccess}
```

Will result in `NULL` because the getter is not used as soon as the object was detected as implementing `\ArrayAccess`.

IMHO correct would be to check if `getSomePropertyNotAccessibleByArrayAccess` (or any of the other allowed "getter" methods) is available or if it is a public property, before using ArrayAccess.

**Comment by bwaidelich** [comment]
Just for reference:
This is the logic that currently applies in the Flow Package:
For `{someThing.foo}`:
* `$someThing->getFoo()`
* `$someThing->isFoo()`
* `$someThing->hasFoo()`
* `$someThing->foo` (if `\stdClass`)
* `$someThing['foo']` (if `\ArrayAccess` or `\SplObjectStorage`

in that order

**Comment by NamelessCoder** [comment]
> This is the logic that currently applies in the Flow Package:

Fluid's logic has been optimised for performance and clarity - and this is what now prevents this use case. The particular behavior from Flow comes from the ObjectAccess (correct me if wrong please!) and can be implemented if desired, using a custom VariableProvider (see https://github.com/NamelessCoder/TYPO3.CMS/blob/master/typo3/sysext/fluid/Classes/Core/Variables/CmsVariableProvider.php for reference; this provider is solely designed to delegate object accessing to ObjectAccess).

I guess my actual question is: how prevalent is this use case of mixing ArrayAccess with getter methods and does that prevalence merit changing the intentionally optimised accessing in Fluid?

**Comment by kitsunet** [comment]
First of all this can not be fixed in the VariableProvider, I tried but due to predetermination of Accessors you can end up with a compiled template directly accessing via:
`...->getVariableProvider()['foo']['bar']` Where `foo` is the array access object and `bar` is a gettable property.

That's also why I think the fix is not hurting performance too much as it still checks on compile time. If the ArrayAccess has no gettable property with the given name the resulting compiled template still uses direct array access, so in the typical case the compiled template will have the same speed as before.

I will come to our example directly, but even the VariableProvider does this.
Eg. the `getAllIdentifiers` method declared in the interface would not be accessible via fluid.

The actual example I stumbled over is our Form framework. It provides a set of nested renderables (eg. Form->Pages->Elements) which are iteratable and accessible through ArrayAcces via the renderable identifier. So each renderable provides the children. But each renderable also has properties itself obviously that need to be accessed. This is now no longer possible.

**Comment by bwaidelich** [comment]
> The particular behavior from Flow comes from the ObjectAccess

Correct. Although it's more that Fluid drove the behavior of `ObjectAccess` in that case.

> how prevalent is this use case of mixing ArrayAccess with getter methods 

IMO it's a perfectly valid usecase and it's not only getters but also:

```php
class SomeClass implements \ArrayAccess {
  public function isFoo() {
  }
  // ...
}
```

**Comment by kitsunet** [comment]
Fixed for now with #185
